### ISSUE TITLE
New Rule: No unnecessary negation (or no double negatives)

### ISSUE THREAD

**Issue by MrJohz** [eslint/eslint#18588]
### Rule details

Warns in cases where negation is used, but the expression being negated involves an equality (e.g. `===` or `<`) that can be inverted.

### Related ECMAScript feature

equalities and negation

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// should be `a !== b`
if (!(a === b)) {
    doSomething();
}

// should be `a >= b`
if (!(a < b)) {
    doSomething();
}

// boolean conditions assigned to variables or used in
// non-if contexts should also get checked
const isValid = !(a !== b);

// complex boolean conditions should get checked
if (condA && x === y && !(a >= b)) {
    doSomething();
}

// negating a negation should also be warned against if the
// expression being negated is an equality (and therefore must
// return a boolean value in the first place)
// should be `a === b` and `a > b` respectively
if (!(!(a === b))) {
    doSomething();
} else if (!!(a > b)) {
    doSomething
}

// This sort of case would **not** be covered, as any attempt at simplifying
// it would be subjective
if (!(a === b || c === d)) {
    doSomething();
}
```

### Why should this rule be in the core instead of a plugin?

This is a general Javascript rule aimed at simplifying expressions that can trivially be simplified.  If-statements with conditionals in are a part of everyday Javascript, and simplifying expressions like this should always produce more readable code.

Specifically, I've seen this problem come up when applying other processes that simplify conditional statements down.  Often, the developer wrote a more complicated conditional statement (e.g. `!(a === b || c === d)`).  Somehow (e.g. through refactoring or by using type-driven lints using typescript-eslint), they figure out that they can remove the `c === d` condition.  This results in `!(a === b)`, which they may well forget to rewrite to the simplified form.

So this lint is designed to catch those cases where there is a trivial simplified form that a developer has forgotten about.

The only case where I can see this causing problems is with NaNs.  If `a` is `NaN`, then `!(a !== a)` is true, but `a === a` is false.  Checking `a !== a` is a semi-idiomatic way of checking that a value is NaN, so this could create false positives.

To resolve this, the lint should specifically not warn if the conditional is `===` and the operands are identical.  (Possibly this could be configurable - if you're consistently using `Number.isNaN` in your codebase, this shouldn't be relevant.)

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

I did think about whether this is just an extension of an existing rule, but I don't think that's the case.

* It is related to the `no-negated-condition` rule, but that specifically targets if-statements, and does not try to simplify the expression at all.  In the situation where `!(a === b)` were converted into `a !== b`, the `no-negated-condition` rule would still apply.
* It may overlap partially with `no-extra-boolean-cast`.  Again, that specifically targets if-statements and other situations where a boolean cast already occurs implicitly, whereas this is about simplifying equalities in general.  However, `if(!!(a > b))` would technically be covered by both rules -- `no-extra-boolean-cast` would remove the boolean cast because the cast already occurs implicitly, whereas this rule would remove it because the case isn't necessary in the first place.  I'm not sure if this is an issue.

However, even where there are overlaps, I think this is distinct from both existing rules, and worth implementing as its own rule.

**Comment by mdjermanovic** [comment]
Hi @MrJohz, thanks for the issue!

We're not accepting new stylistic rules in the core unless they relate to new ECMAScript features that reached stage 4 in the preceding 12 months. For more details, please see [this blog post](https://eslint.org/blog/2020/05/changes-to-rules-policies/).

I think the exact rule you're looking for already exists in `eslint-plugin-sonarjs`:

https://github.com/SonarSource/eslint-plugin-sonarjs/blob/HEAD/docs/rules/no-inverted-boolean-check.md

Although the example shows only `===`, I checked it and it reports all cases from your example code.

![image](https://github.com/eslint/eslint/assets/44349756/1747d01a-866e-496d-8613-139916d46ace)
